c7n/resources/vpc.py

```diff
@@ -3287,3 +3287,45 @@ def process(self, resources, event=None):
                 results.append(resource)
 
         return results
+
+
+@NetworkAddress.filter_registry.register('used-by')
+class UsedByNetworkAddress(Filter):
+    """Filter Elastic IPs to find the resource type that the network
+    interface that the Elastic IP is associated with is attached to.
+
+    This filter is useful for limiting the types of resources to
+    enable AWS Shield Advanced protection.
+
+    :Example:
+
+    .. code-block:: yaml
+
+            policies:
+              - name: eip-shield-advanced-enable
+                resource: aws.elastic-ip
+                filters:
+                  - type: used-by
+                    resource-type: elb-net
+                  - type: shield-enabled
+                    state: false
+                actions:
+                  - type: set-shield
+                    state: true
+    """
+    schema = type_schema(
+        'used-by', required=['resource-type'], **{
+            'resource-type': {'type': 'string'}}
+    )
+    permissions = ("ec2:DescribeNetworkInterfaces",)
+
+    def process(self, resources, event=None):
+        self.nics = self.manager.get_resource_manager('eni').resources()
```

Comment:
I don't think we need an instance variable here, and ideally we'd use get_resources to fetch nics for the eips, some accounts have *lots* of them.

Reply:
@kapilt Thanks for catching it! Regarding `get_resource_manager`, I am using the same logic [here](https://github.com/cloud-custodian/cloud-custodian/blob/main/c7n/resources/vpc.py#L1013-L1019). We heavily use the SG `used`/`unused` filters in our environment and haven't run into any issues.

Reply:
we always want to default to efficient api usage, sg used/unused do to need to process all enis. but in this case we know exactly which ends we want. ie. if we have 10 elastic ips, and a million enis, we want to just fetch the 10 enis we want to look at versus trying to get the whole population.

```
enis = get_resource_manager('eni').get_resources(set_of_enis_ids_from_eip)
```

Reply:
@kapilt Thanks for your guidance! I updated the code as you suggested.